<div style="width: 100%; clear: both;">
    <div style="float: left; width: 50%;">
       <img src="http://www.uoc.edu/portal/_resources/common/imatges/marca_UOC/UOC_Masterbrand.jpg", align="left">
    </div>
</div>

<div style="float: right; width: 50%;">
    <p style="margin: 0; padding-top: 22px; text-align:right;">M1.871 · Privacidad</p>
    <p style="margin: 0; text-align:right;">Máster Universitario en Ciberseguridad y Privacidad</p>
    <p style="margin: 0; text-align:right; padding-button: 100px;">Estudios de Informática, Multimedia y Telecomunicación</p>
</div>

</div>
<div style="width: 100%; clear: both;">
<div style="width:100%;">&nbsp;</div>

# Práctica 1

--- 

Para resolver esta práctica, tenéis que poner las soluciones en el mismo
*notebook*, en las celdas de código que están en blanco habilitadas para ello. Después de cada celda de código podéis poner una celda *markdown* en con la justificación de la respuesta dada en la solución. También 
podéis poner justificaciones en forma de comentario en las celdas de código. No es obligatorio, pero sí muy recomendable que justifiquéis las respuestas que dais, ya que eso facilita la corrección en caso de una respuesta incorrecta (pudiendo valorar el planteamiento).

En muchos casos, cuando se pide implementar una función, se proporciona el nombre de la función y los parámetros, o algún tipo de esqueleto, indicando como ayuda el tipo de dichos parámetros y el tipo que retorna la función en forma de *type hint*. Es necesario que respetéis el nombre de la función y en la medida de los posibles los tipos. Sin embargo podéis implementar el código de la función con total libertad. Podéis
definir otras funciones adicionales y no respetar del todo los *type hints* si así lo consideráis oportuno. 

Este notebook va acompañado de un cuestionario online con preguntas sobre los ejercicios. Este cuestionario debe ser contestado antes de la fecha de etrega de la práctica. **Se recomienda que afrontéis el cuestionario una vez finalizada la practica, habiendo realizado y entendido todos los ejercicios.** En muchos casos se proporcionan datos o posibles resultados con el objeto de ayudar a poder contestar el cuestionario aunque no se haya acabado o no se haya sabido implementar todos los ejercicios de este *notebook*.

En algunas preguntas hay algunas indicaciones finales como *Hints*, son sugerencias sobre el uso de funciones que os pueden ser utiles. En cualquier caso se trata simplemente de eso, sugerencias.

# Datos

Trabajaremos con datos que simulan el expediente academico de alumnos. Tenemos el fichero `data/students1.csv` que contiene 10000 registros con los siguientes atributos:
- `name`: nombre del alumno/a,
- `subject0`: nota de la asignatura 0 (valor real entre 0 y 10),
- `subject1`: nota de la asignatura 1,
- `subject2`: nota de la asignatura 2,
- `subject3`: nota de la asignatura 3,

No podemos publicar este fichero de microdatos, ya que se considera que la nota obtenida por el alumnado es privada. Es decir no podemos saber que nota ha obtenido una estudiante en concreto.



# Publicación de microdatos

Queremos ahora poder publicar una versión protegida del fichero de microdatos `data/students1.csv`. Para poder trabajar con estos datos utilizaremos la libreira `pandas` de Python. De esta manera podemos leer el fichero en memoria:

In [1]:
import numpy as np
import pandas as pd

df1 = pd.read_csv("data/students1.csv")
df1.info()

df1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   name      10000 non-null  object 
 1   subject0  10000 non-null  float64
 2   subject1  10000 non-null  float64
 3   subject2  10000 non-null  float64
 4   subject3  10000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 390.8+ KB


name  subject0  subject1  subject2  subject3
0          Susan Duffy       8.5       6.1       1.0       8.8
1           Erin Price       0.1       4.9       1.8       5.2
2        Harvey Warren       2.5       7.0       4.1       9.7
3      Russell Sanchez       6.5       6.4       6.1       9.0
4         John Bohrman       0.8       9.7       2.3       5.3
...                ...       ...       ...       ...       ...
9995  Wilbert Meredith       0.3       1.8       6.1       9.8
9996     Thomas Gordon       4.9       5.8      10.0       3.6
9997     Kathryn Lakey       0.2      10.0       3.7       6.5
9998     Dewey Miesner       8.8       8.5       1.5       5.2
9999   Susan Tornatore       0.8       2.4       5.7       7.2

[10000 rows x 5 columns]

Como vemos, ahora tenemos en la variable `df1` la tabla de microdatos en una estructura [`DataFrame`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) de `pandas`. Notad que tamibén hemos importado la libreria `numpy`, ya que la utilizaremos más adelante.

## Ejercicio 1 
*[0.5p]*

Para publicar estos datos primero tenemos que ver si se pueden publicar todos los atributos.

En el caso de que haya algún atributo que no se pueda publicar, elimínalo del `DataFrame` `df1`.

*Hint*:
- [pandas.DataFrame.drop](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html)


In [2]:
## Eliminaremos la columna name, que permitiría identificar totalmente el estudiante.

df1 = df1.drop(columns=['name'])
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   subject0  10000 non-null  float64
 1   subject1  10000 non-null  float64
 2   subject2  10000 non-null  float64
 3   subject3  10000 non-null  float64
dtypes: float64(4)
memory usage: 312.6 KB


# Ejercicio 2 
*[1.25p]*

Para el resto de atributos vamos a utilizar **ruido aditivo no correlacionado**, tal como se describe en los materiales. 

Para ello vamos a utilizar una distribución normal $N(\mu, \sigma^2)$ con $\mu = 0$, y como varianza tomaremos $\sigma^2 = p \sigma_o^2$
donde $\sigma_o^2$ es la varianza de la variable que queremos proteger.

Desarrollad una función `noise_add_normal(df, p)` que recibe como parámetro `df`, el DataFrame a anonimizar y el parámetro `p`, y retorna una copia del DataFrame protegido, aplicando ruido a cada atributo. 

Finalmente, obtened una versión protegida en la variable `df1_noise_1` utilizando $p=0.1$.

Algunas consideraciones:
- cada atributo (variable o columna) tendrá una distribución de ruido independiente.
- la función no puede modificar el DataFrame original, tiene que retornar un DataFrame nuevo.
- tened cuidado a la hora de pasar la varianza a la distribución normal de la librería `numpy`, fijaros que espera recibir la desviación estándar (no la varianza).
- el resultado después de aplicar ruido NO tiene que estar redondeado a un decimal (como sí lo estaban los valores originales), ni se tiene que aplicar ninguna corrección.


*Hint*:
- [`numpy.random.Generator.normal`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.normal.html#numpy.random.Generator.normal)
- [Random sampling, quick start](https://numpy.org/doc/stable/reference/random/index.html#random-quick-start)


In [3]:
import math

def noise_add_normal(df: pd.DataFrame, p: float) -> pd.DataFrame:
    ## Code goes here
    rng = np.random.default_rng()

    result = pd.DataFrame(columns=df.columns)
    for name, values in df.items():
        N = values.size
        variance = values.var()
        sigma = math.sqrt(p*variance)
        mu = 0
        ## https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.normal.html
        vector = rng.normal(mu, sigma, N)
        result[name] = values + vector

    return result

## Code goes here
df1_noise_1 = noise_add_normal(df1, 0.1)
#df1_noise_1.info()
print (df1_noise_1)

for name, values in df1.items():
    print (name)
    print (f"      media antes={values.mean()} despues={df1_noise_1[name].mean()}")
    print (f"  variancia antes={values.var()} despues={df1_noise_1[name].var()}")


      subject0   subject1  subject2   subject3
0     9.715860   7.237924  0.315162   8.258443
1    -0.001542   5.618070  3.548728   7.079863
2     2.941535   5.132987  6.068056   9.330093
3     6.298366   6.697849  6.857184   9.097682
4     1.310652   9.452741  1.506636   5.108540
...        ...        ...       ...        ...
9995  0.351684   0.360472  8.834801  10.783407
9996  5.754225   5.448386  9.245939   5.072592
9997  2.938843  12.460461  3.231329   6.557660
9998  6.562335   8.982930  1.174824   5.988478
9999  2.063235   0.647393  4.831024   6.024327

[10000 rows x 4 columns]
subject0
      media antes=5.41787 despues=5.4324218065011705
  variancia antes=9.737967459845985 despues=10.608356907622232
subject1
      media antes=5.46277 despues=5.452997633687718
  variancia antes=9.33161808890889 despues=10.239441719239597
subject2
      media antes=5.408580000000001 despues=5.4229257478259765
  variancia antes=9.470857469346933 despues=10.39027995994526
subject3
      media antes=5

Como guía, los valores protegidos en `df1_noise_1` pueden ser algo parecido a lo que se muestra en la siguiente tabla.

|  | subject0 |	subject1 |	subject2 |	subject3 |
| --- | --- | --- | --- | --- |
| 0 | 6.626006 |	6.858254 |	3.024407 |	10.113438 |
| 1 | -0.190741 |	5.878539 |	1.641941 |	4.289627 |
| 2 | 0.326832 |	7.113423 |	2.489244 |	10.214147 |
| 3 | 7.815629 |	5.846698 |	7.173413 |	10.008515 |
| 4 | 1.246823 |	10.570642 |	2.149917 |	4.912380 |
| 5 | ... | ... | ... | ... |
| ... | ... | ... | ... | ... |

## Ejercicio 3
*[1.25p]*

En este ejercicio desarrollaremos dos medidas de **pérdida de información**.

Para ello vamos a utilizar la siguiente notación:
- $X$: conjunto de datos original. Lo podemos interpretar como una matriz de
tamaño $n \times d$ donde $n$ es el número de registros (filas) y $d$ el
número de variables (columnas).
- $X'$: conjunto de datos protegido, también $n \times d$.
- $R$: matriz de correlación de los datos originales $X$. Esta matriz es una matriz $d \times d$ donde el elemento $(i,j)$ es la correlación entre la variable $V_i$ y $V_j$. Como medida de correlación tomaremos el coeficiente de correlación de Pearson.
- $R'$: matriz de correlación de los datos protegidos $X'$.

En cada caso mediremos la media del error absoluto MAE (*mean absolute error*). Es decir:
- Sobre $X$, $X'$:
  - *MAE*: $\frac{1}{nd}\sum_{j=1}^d \sum_{i=1}^n \mid x_{ij} - x'_{ij} \mid$
- Sobre $R$, $R'$:
  - *MAE-corr*: $\frac{1}{d^2}\sum_{j=1}^d \sum_{i=1}^n \mid r_{ij} - r'_{ij} \mid$

Se pide la implementación de la función `information_loss(df_original, df_protected)` donde `df_original` es el DataFrame con los datos originales ($X$) y `df_protected` el DataFrame con los datos protegidos ($X'$). La función tiene que retornar dos valores, en este orden:
- valor del MAE sobre $X$ y $X'$
- valor del MAE sobre $R$ y $R'$

Una vez implementada, aplicad la función para medir la pérdida de información al proteger los datos con $p=0.1$ (`df1_noise_1` del ejercicio anterior).



In [4]:
from typing import Tuple

def MAE(df_original: pd.DataFrame, df_protected: pd.DataFrame) -> float:
    d = df_original.columns.size
    N = df_original.size

    ## MAE sobre los datos
    suma = 0.0
    for name, values in df_original.items():
        tmp = abs(values - df_protected[name])
        suma = suma + sum(tmp)
    #print (suma)
    return suma  / (N * d)

def MAE_matrix(original: np.array, protected: np.array) -> float:
    d = len(original)
    tmp = abs(original - protected)
    #print (tmp)    
    suma = sum(sum(tmp))
    #print (suma)
    return suma  / (d * d)

def matriz_covarianza (df_original: pd.DataFrame) -> np.array:
    ## calculamos la matriz de covarianza cov(Vi,Vj)
    d = df_original.columns.size
    N = df_original.size
    covarianza = [[0 for _ in range(d)] for _ in range(d)]
    for idx1, name1 in enumerate(df_original.columns):
        values1 = df_original[name1]
        media1 = values1.mean()
        for idx2, name2 in enumerate(df_original.columns):
            values2 = df_original[name2]
            media2 = values2.mean()
            if name1 != name2:
                covij = sum((values1 - media1)*(values2 - media2)) / N
                #print (name1 + "-" + name2 + " " + str(covij))
                covarianza[idx1][idx2] = covij
            else:
                covarianza[idx1][idx2] = 0.0
    #print (covarianza)
    return np.array(covarianza)

def matriz_correlacion (df_original: pd.DataFrame) -> np.array:
    ## calculamos la matriz de covarianza cov(Vi,Vj) y la dividimos por la desviacion estandar
    d = df_original.columns.size
    N = df_original.size
    covarianza = [[0 for _ in range(d)] for _ in range(d)]
    for idx1, name1 in enumerate(df_original.columns):
        values1 = df_original[name1]
        stddev1 = math.sqrt(values1.var())
        media1 = values1.mean()
        for idx2, name2 in enumerate(df_original.columns):
            values2 = df_original[name2]
            stddev2 = math.sqrt(values2.var())
            media2 = values2.mean()
            if name1 != name2:
                covij = sum((values1 - media1)*(values2 - media2)) / N
                #print (name1 + "-" + name2 + " " + str(covij))
                # coeficiente de correlación de Pearson
                covarianza[idx1][idx2] = covij / (stddev1 * stddev2) 
            else:
                covarianza[idx1][idx2] = 0.0
    #print (covarianza)
    return np.array(covarianza)

def information_loss(df_original: pd.DataFrame, df_protected: pd.DataFrame) -> Tuple[float, float]:
    
    ## calculamos el MAE (mean absolute error) entre la original y la modificada
    mae = MAE(df_original, df_protected)

    ## calculo de las matrices de correlación

    ## calculamos la matriz de correlacion del original cov(Vi,Vj)/std*std
    correlacion_original = matriz_correlacion (df_original)
    ## calculamos la matriz de correlacion del modificado cov(Vi,Vj)/std*std
    correlacion_protected = matriz_correlacion (df_protected)

    ## calculamos el MAE (mean absolute error) entre las matirzes de correlacion
    mae_corr = MAE_matrix(correlacion_original, correlacion_protected)

    return [mae, mae_corr]

## Code goes here
print (f"Perdida de información {information_loss(df1, df1_noise_1)}")

Perdida de información [0.19524913058322446, 0.000803730756544534]


## Ejercicio 4
*[2p]*

Queremos ahora producir una versión de los microdatos con el mayor nivel de ruido posible pero sin perder mucha información. Concretamente, queremos garantizar que el MAE (unicamente el MAE, no el MAE-corr) no supere el valor $3$.

Es decir, ¿qué nivel de ruido (parámetro $p$) podemos aplicar de manera que obtengamos el mayor nivel de privacidad pero manteniendo la perdida de información según el MAE por debajo de $3.0$?

Se pide que programéis el código para determinar dicho valor $p$. Para restringir la búsqueda, consideraremos que el parámetro $p$ puede tomar valores a partir de $0$ y siempre se expresará con un solo decimal. Es decir, podemos tener valores como $p=0.3$ o $p=1.2$ pero NO $p=0.31$ o $p=1.023$.

Tenéis que mostrar al final el valor de $p$ que buscamos y el MAE obtenido con dicho valor.


In [5]:
## Code goes here
p = 0.0
found = False

while not found:
    #print (str(p))
    df1_noise_2 = noise_add_normal(df1, p)
    mae_pairs = information_loss(df1, df1_noise_2)
    
    if (mae_pairs[0] > 3.0):
        found = True
    else:
        #print (mae_pairs)
        df1_noise_2_found = df1_noise_2
        mae_pairs_found = mae_pairs
        p_found = p
    p = p + 0.1

if found:
    print (f"Valor p utilizado {p_found} con MAE {mae_pairs_found[0]}")    
else:
    print ("No he encontrado ningún valor")    

Valor p utilizado 23.400000000000063 con MAE 2.9736782689654744


## Ejercicio 5
*[1.25p]*

Con el valor $p$ obtenido en el ejercicio anterior, vamos a obtener un nuevo DataFrame protegido:
```Python
df_protected_final = noise_add_normal(df1, p)
```
(Si no has podido acabar el ejercicio anterior, utiliza el valor $p=1.8$, que no tiene por qué ser igual a la solución de dicho ejercicio)

Este será el fichero de microdatos publicado. Para ver que diferencias tenemos respecto al original se pide comparar los siguientes valores entre el fichero de notas original (`df1`) y el protegido (`df_protected_final`):
- media de cada asignatura por separado.
- media de todas las asignaturas.
- nota máxima y mínima de cada asignatura por separado
- nota máxima y mínima de todas las asignaturas.

Para ello, tenéis que implementar una función `compute_statistics` que reciba como parámetro un `DataFrame` con las notas y retorne un diccionario con los valores requeridos. Como ejemplo, si aplicamos la función a los datos originales, `df1`, debería retornar:
```python
{'avg_s0': 5.4178700000000015,
 'avg_s1': 5.462769999999979,
 'avg_s2': 5.408580000000002,
 'avg_s3': 5.327550000000016,
 'avg_total': 5.4041925,
 'max_s0': 10.0,
 'max_s1': 10.0,
 'max_s2': 10.0,
 'max_s3': 10.0,
 'max_total': 10.0,
 'min_s0': 0.0,
 'min_s1': 0.0,
 'min_s2': 0.0,
 'min_s3': 0.0,
 'min_total': 0.0}
```
De esta manera podemos comparar los valores sobre `df1` y `df_protected_final`
```python
stats = pd.DataFrame([compute_statistics(df1), compute_statistics(df_protected_final)])

```
En `stats` tenemos un `DataFrame` donde podemos ver los resultados.
`
En este ejercicio se pide:
- implementar la función `compute_statistics` tal como se pide,
- mostrar el resultado en el `DataFrame` stats como se indica anteriormente.

In [6]:
from typing import Dict

def compute_statistics(df: pd.DataFrame) -> Dict[str, float]:
    ## Code goes here
    result = dict()
    n = 0
    sum = 0.0
    min_v = 0
    max_v = 0
    for name, values in df.items():
        value = values.mean()
        result[f"avg_{name}"] = value
        result[f"max_{name}"] = values.max()
        result[f"min_{name}"] = values.min()
        if n > 0:
            min_v = min(min_v, values.min())
            max_v = max(max_v, values.max())
        n = n + 1
        sum = sum + value

    result[f"avg_total"] = sum / n
    result[f"max_total"] = max_v
    result[f"min_total"] = min_v
    #print (result)
    return result

## Code goes here
df_protected_final = noise_add_normal(df1, p_found)

stats = pd.DataFrame([compute_statistics(df1), compute_statistics(df_protected_final)])
print (f"dataframe con stats {stats}")

dataframe con stats    avg_subject0  max_subject0  min_subject0  avg_subject1  max_subject1  \
0      5.417870     10.000000      0.000000      5.462770     10.000000   
1      5.219324     57.964143    -54.317962      5.560443     59.141741   

   min_subject1  avg_subject2  max_subject2  min_subject2  avg_subject3  \
0      0.000000      5.408580     10.000000      0.000000      5.327550   
1    -53.602085      5.496744     66.411919    -48.325699      5.291929   

   max_subject3  min_subject3  avg_total  max_total  min_total  
0     10.000000      0.000000   5.404193  10.000000   0.000000  
1     60.296152    -53.760474   5.392110  66.411919 -53.760474  


# Privacidad diferencial

Ahora vamos a ver el enfoque que propone la privacidad diferencial, y para ello implementaremos de forma simulada una base de datos como una clase de Python.


La siguiente clase implementa una sencilla base de datos `MyDB`. La base de datos lee el fichero CSV que contiene los datos con las notas de los estudiantes.

Como vemos, se trata de una base de datos muy sencilla y simplificada, que únicamente tiene sentido con fines docentes. La idea es que la clase `MyDB` simula una base de datos y sus métodos públicos son las consultas que pueden hacer los usuarios. Asumimos que un usuario no puede ver el contenido del objeto y simplemente puede acceder a `MyDB` mediante sus métodos.

La clase `MyDB` se instancia a partir de un fichero CSV que se guarda como un `DataFrame` de `pandas`. Esta clase implementa 1 método para obtener información sobre las estudiantes de la base de datos:
- `count_all_fail`: que retorna el número de estudiantes que han suspendido todas las asignaturas.

Recordamos que asumimos que la única manera que tiene cualquier usuario para acceder a los datos es llamar a esta función `count_all_fail`.


In [7]:

class MyDB:
    def __init__(self, filename: str) -> None:
        self.df = pd.read_csv(filename)

    def count_all_fail(self) -> int:
        return self.df[
            (self.df["subject0"] < 5)
            & (self.df["subject1"] < 5)
            & (self.df["subject2"] < 5)
            & (self.df["subject3"] < 5)
            ].shape[0]


Como ejemplo, a continuación vemos como se instanciaría `SimpleDB` con el fichero `data/students1.csv`

In [8]:
db1 = MyDB("data/students1.csv")

print(f"DB1 count_all_fail: {db1.count_all_fail()}")

DB1 count_all_fail: 420


## Ejercicio 6
*[0.5p]*

Vamos ahora a estudiar la privacidad que ofrece `MyDB` según la definición de privacidad diferencial. Con el objeto de poder hacer diversas pruebas tenemos dos ficheros CSV:
- `data/students1.csv`: con 10000 registros
- `data/students2.csv`: con 9999 registros

Podemos ver que `students2.csv` es igual a `students1.csv` con la diferencia de que le falta un registro, el correspondiente a *Susan Duffy*. En principio podemos asumir que un posible atacante sabe que este es el registro que falta (el correspondiente a esta estudiante no está en DB2 y sí en DB1), pero no sabe sus datos: notas obtenidas en las asignaturas.

In [9]:
db1 = MyDB("data/students1.csv")
db2 = MyDB("data/students2.csv")


Observando el resultado de hacer la consulta `count_all_fail` a `db1` y `db2`. ¿Qué información podemos obtener? (Muestra el resultado de las consultas y añade la respuesta y su justificación como comentario).


In [10]:
## Code goes here
print (f"número de suspendidos con Susan {db1.count_all_fail()} y sin Susan {db2.count_all_fail()}")

# Si el número de suspendidos de las dos muestras es igual es que el estudiante eliminado no suspendió
if db1.count_all_fail() == db2.count_all_fail():
    print ("Susan no ha suspendido la asignatura")
else:
    print ("Susan ha suspendido la asignatura")


número de suspendidos con Susan 420 y sin Susan 420
Susan no ha suspendido la asignatura


## Ejercicio 7
*[1.25p]*

Vamos a implementar una nueva base de datos que implemente la versión segura del método `count_all_fail` utilizando el mecanismo de Laplace. Para ello crearemos una nueva clase de base de datos a la que llamaremos `MyPrivateDB`.



Implementa, en la siguiente clase, el método `dp_count_all_fail` que retorne el número de alumnos y alumnas que han suspendido todas las asignaturas, pero esta vez la respuesta ha de estar protegida utilizando **el mecanismo de Laplace**.

Para hacerlo necesitaremos:
- Calcular o determinar la sensibilidad global del método.
- Generar ruido con una distribución de Laplace basándonos en su sensibilidad global y el parámetro *epsilon*.
- Retornar el valor con el ruido añadido.

Como vemos, el parámetro *epsilon* se fija en el constructor de la clase. Importante: de momento asumimos que únicamente se realizará una consulta.

*Hint*: 
- [random.Generator.laplace](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.laplace.html#numpy.random.Generator.laplace)

In [11]:
import numpy as np

class MyPrivateDB:
    def __init__(self, filename: str, epsilon: float) -> None:
        self.df = pd.read_csv(filename)
        self.epsilon = epsilon
        self.rng = np.random.default_rng()
        # en este caso la sensibilidad global es 1, porque la diferencia máxima entre el número de estudiantes suspendidos 
        # entre la muestra con y sin un estudiante, sólo puede ser de 1, dado que o suspende o no.
        self.global_sensitivity = 1

    def dp_count_all_fail(self) -> int:
        ## Code goes here
        original_value = self.df[
            (self.df["subject0"] < 5)
            & (self.df["subject1"] < 5)
            & (self.df["subject2"] < 5)
            & (self.df["subject3"] < 5)
            ].shape[0]

        # aplicamos ruido al número de alumnos suspendidos, mediante la distribución de Laplace centrada en 0 y escala
        # calculada a partir de la sensibilidad global y el epsilon. 
        lap = self.rng.laplace(0, self.global_sensitivity / self.epsilon, 1)
        return int(original_value + lap[0])


## Ejercicio 8
*[0.5p]*

Podemos ver el resultado de hacer la consulta `dp_count_all_fail` con la nueva base de datos `MyPrivateDB` a los datos de `data/students1.csv` y `data/students2.csv`.

¿Podemos ahora obtener la misma información que en el Ejercicio 6? Muestra el resultado para `epsilon=1`, responde y justifica en forma de comentario.

In [12]:
## Code goes here

db1_private = MyPrivateDB("data/students1.csv", epsilon=1)
db2_private = MyPrivateDB("data/students2.csv", epsilon=1)

print (f"número de suspendidos con Susan {db1_private.dp_count_all_fail()} y sin Susan {db2_private.dp_count_all_fail()}")
# dado que el número de suspendidos entre las dos muestras, está "manipulado", con un primer resultado no podemos determinar a ciencia cierta si 
# Susan está o no suspendida

número de suspendidos con Susan 418 y sin Susan 419


## Ejercicio 9
*[2p]*

Hemos asumido al principio que el usuario o usuaria únicamente realizará una consulta (llamará a la función `dp_count_all_fail()` una vez).

Si ahora asumimos que se puede hacer más de una consulta, y sin modificar la implementación de nuestro método `dp_count_all_fail`, un atacante ¿podría llegar a obtener la misma información que proporcionaba el método `count_all_fail` pero utilizando `dp_count_all_fail`?

En este ejercicio se pide implementar el posible ataque que permita desanonimizar `dp_count_all_fail` (es decir, utilizando `MyPrivateDB`) y con la particularidad de que se puede repetir dicha consulta sobre una misma base de datos diversas veces. En este caso lo haremos para `epsilon=8`. Justifica el resultado.

In [13]:
## Code goes here
## supondremos que students2.csv contiene los datos de ayer y 
## students1.csv los datos de hoy, y es donde se incorporan los datos de Susana de la que queremos descubrir si ha suspendido o no

# contiene los datos de ayer
db2_private8 = MyPrivateDB("data/students2.csv", epsilon=8)


# contiene los datos de hoy
db1_private8 = MyPrivateDB("data/students1.csv", epsilon=8)

# tenenmos el resultado de la consulta de ayer
suspendidos_ayer = db2_private8.dp_count_all_fail()

# como la consulta resultante, está modificada, realizaremos varios intentos, para ver como se comporta la diferencia
n = 0
s = 0
for i in range(80):
    # realizamos la consulta "hoy"
    suspendidos_hoy = db1_private8.dp_count_all_fail()
    if suspendidos_hoy <= suspendidos_ayer:
        s = s + 1
    n = n + 1

print (f"después de {n} tests hay un total de {s} muestras en que no hay mas suspendidos entre ayer y hoy. Eso quiere decir que hay una probabilidad del {100.0*s/n}% de que Susana no haya suspendido")



después de 80 tests hay un total de 43 muestras en que no hay mas suspendidos entre ayer y hoy. Eso quiere decir que hay una probabilidad del 53.75% de que Susana no haya suspendido
